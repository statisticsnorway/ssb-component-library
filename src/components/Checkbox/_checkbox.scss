.ssb-checkbox {
  @include roboto;
  color: $ssb-dark-6;
  cursor: pointer;
  display: block;
  font-size: 16px;
  min-width: 200px;
  outline: none;
  position: relative;
  text-align: left;
  transition: background .18s;
  user-select: none;

  &:hover, &:focus {
    background: $ssb-green-2;

    .checkbox-label {

      &:before {
        border: 2px solid $ssb-green-4;
      }
    }
  }

  input[type=checkbox] {
    cursor: pointer;
    opacity: 0;
    position: absolute;

    &:checked + .checkbox-label {

      &:before {
        background: $ssb-dark-5;
      }

      &:after {
        border: solid $ssb-white;
        border-width: 0 3px 3px 0;
        content: '';
        display: block;
        height: 12px;
        left: 17px;
        position: absolute;
        top: 14px;
        transform: rotate(45deg);
        width: 6px;
      }
    }

    &:disabled + .checkbox-label {
      color: $ssb-dark-2;

      &:before {
        border: 1px solid $ssb-dark-2;
      }
    }
  }

  .checkbox-label {
    align-items: center;
    cursor: pointer;
    display: flex;
    margin-bottom: 0;
    padding: $global-padding-size;
    position: relative;

    &:before {
      background: $ssb-white;
      border: 1px solid $ssb-dark-5;
      border-radius: 2px;
      content: '';
      display: block;
      height: 20px;
      margin-right: $global-padding-size;
      width: 20px;
    }
  }
}
