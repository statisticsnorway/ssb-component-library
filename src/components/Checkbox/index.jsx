import React from 'react';
import PropTypes from 'prop-types';

const Checkbox = ({
	index, children, callback, disabled, selected, value,
}) => (
	<div className="ssb-checkbox" tabIndex={index + 1}>
		<input
			disabled={disabled}
			id={value}
			checked={selected}
			onChange={() => callback(value)}
			type="checkbox"
			value={value}
		/>
		<label className="checkbox-label" htmlFor={value}>{children}</label>
	</div>
);

Checkbox.defaultProps = {
	callback: () => {},
	index: 1,
};

Checkbox.propTypes = {
	callback: PropTypes.func,
	children: PropTypes.node,
	disabled: PropTypes.bool,
	index: PropTypes.number,
	selected: PropTypes.bool,
	value: PropTypes.oneOfType([
		PropTypes.string,
		PropTypes.number,
	]).isRequired,
};

export default Checkbox;
