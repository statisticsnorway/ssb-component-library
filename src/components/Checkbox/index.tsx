import React, { type ReactNode } from 'react'

interface CheckboxProps {
  callback?: (value: string | number) => void
  children?: ReactNode
  className?: string
  disabled?: boolean
  selected?: boolean
  tabIndex?: number
  value: string | number
}

const Checkbox: React.FC<CheckboxProps> = ({
  callback = () => {},
  children,
  className,
  disabled,
  selected,
  tabIndex = 0,
  value,
}) => (
  <div className={`ssb-checkbox${className ? ` ${className}` : ''}`}>
    <input
      tabIndex={tabIndex}
      disabled={disabled}
      id={value.toString()}
      checked={selected}
      onChange={() => callback(value)}
      type='checkbox'
      value={value}
    />
    <label className='checkbox-label' htmlFor={value.toString()}>
      {children}
    </label>
  </div>
)

export default Checkbox
