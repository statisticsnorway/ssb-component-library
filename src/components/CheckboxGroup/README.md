CheckboxGroup
========

> A wrapper for multiple checkbox buttons

### Usage

#### HTML

```html
<div class="ssb-checkbox-group">
  <div class="ssb-title no-margin"><h5>Title</h5></div>
  <div class="ssb-checkbox">
    <input type="checkbox" id="option1" value="option1" />
    <label class="checkbox-label" for="option1">Insert label</label>
  </div>
  <div class="ssb-checkbox">
    <input type="checkbox" id="option2" value="option2" />
    <label class="checkbox-label" for="option2">Insert label</label>
  </div>
</div>
```

#### React

```jsx harmony
<CheckboxGroup
  header="Header"
  items={
    {
      label: 'Item 1',
      value: 'item1',
    },
    {
      label: 'Item 2',
      value: 'item2',
    },
  }
  onChange={() => callback}
  selectedValue="item2"
/>
```

Available props:

| Name | Type | Description |
| ---- | ---- | ---- |
| header | string | Renders a h5 title |
| items | arrayOf(label, value) | Required items for rendering radio buttons |
| onChange | func | Callback function when a value is changed |
| selectedValue | string | Pre selected value |
