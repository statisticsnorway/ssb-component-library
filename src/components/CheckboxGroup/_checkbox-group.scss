.ssb-checkbox-group {
  @include roboto;
  display: inline-block;

  .checkbox-group-header {
    padding-left: 10px;
  }
}
