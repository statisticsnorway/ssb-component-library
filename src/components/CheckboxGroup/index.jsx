import React, { useEffect, useState } from 'react';
import PropTypes from 'prop-types';
import Checkbox from '../Checkbox';
import Title from '../Title';

const CheckboxGroup = ({
	header, items, onChange, selectedValues,
}) => {
	const [selected, updateSelected] = useState(selectedValues);

	useEffect(() => {
		onChange(selected);
	}, [selected]);

	const setSelected = sel => {
		const newArr = [...selected];
		if (selected.includes(sel)) {
			newArr.splice(selected.indexOf(sel), 1);
		} else {
			newArr.push(sel);
		}
		updateSelected(newArr);
	};

	return (
		<div className="ssb-checkbox-group">
			{header && <Title className="checkbox-group-header no-margin" size={5}>{header}</Title>}
			{items.map((it, index) => (
				<Checkbox
					key={it.value}
					index={index}
					selected={selected.includes(it.value)}
					value={it.value}
					callback={setSelected}
					disabled={it.disabled}
				>{it.label}
				</Checkbox>
			))}
		</div>
	);
};

CheckboxGroup.defaultProps = {
	onChange: () => {},
	selectedValues: [''],
};

CheckboxGroup.propTypes = {
	header: PropTypes.string,
	items: PropTypes.arrayOf(PropTypes.shape({
		label: PropTypes.string,
		value: PropTypes.string,
	})).isRequired,
	onChange: PropTypes.func,
	selectedValues: PropTypes.array,
};

export default CheckboxGroup;
