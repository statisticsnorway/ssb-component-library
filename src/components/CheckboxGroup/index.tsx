import React, { useEffect, useState } from 'react'
import Checkbox from '../Checkbox'
import classNames from '../../utils/utils'

interface Item {
  label: string
  value: string
  disabled?: boolean
}

interface CheckboxGroupProps {
  className?: string
  header?: string
  items: Item[]
  onChange?: (selected: string[]) => void
  orientation?: 'column' | 'row'
  selectedValues?: string[]
}

const CheckboxGroup: React.FC<CheckboxGroupProps> = ({
  className,
  header,
  items,
  onChange = () => {},
  orientation = 'column',
  selectedValues = [''],
}) => {
  const [selected, updateSelected] = useState(selectedValues)

  useEffect(() => {
    onChange(selected)
  }, [selected])

  const setSelected = (sel: string) => {
    const newArr = [...selected]
    if (selected.includes(sel)) {
      newArr.splice(selected.indexOf(sel), 1)
    } else {
      newArr.push(sel)
    }
    updateSelected(newArr)
  }

  return (
    <div className={classNames('ssb-checkbox-group', className)}>
      {header && <div className='checkbox-group-header'>{header}</div>}
      <div className={`boxes flex-${orientation}`}>
        {items.map((it) => (
          <Checkbox
            key={it.value}
            selected={selected.includes(it.value)}
            value={it.value}
            callback={setSelected}
            disabled={it.disabled}
          >
            {it.label}
          </Checkbox>
        ))}
      </div>
    </div>
  )
}

export default CheckboxGroup
