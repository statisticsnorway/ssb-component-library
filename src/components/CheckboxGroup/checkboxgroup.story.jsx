import React from 'react';
import { storiesOf } from '@storybook/react';
import centered from '@storybook/addon-centered/react';
import CheckboxGroup from './index';

const checkboxItems = [
	{
		label: 'Item 1',
		value: 'item1',
	}, {
		label: 'Item 2',
		value: 'item2',
	}, {
		label: 'Item 3',
		value: 'item3',
		disabled: 'true',
	},
];

storiesOf('CheckboxGroup', module).addDecorator(centered)
	.add('Column', () => (
		<CheckboxGroup header="Header" items={checkboxItems} />
	))
	.add('Row', () => (
		<CheckboxGroup header="Header" items={checkboxItems} orientation="row" />
	));
