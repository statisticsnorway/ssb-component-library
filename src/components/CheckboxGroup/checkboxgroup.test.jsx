import React from 'react';
import {mount, shallow} from 'enzyme';
import CheckboxGroup from './index';
import Checkbox from '../Checkbox';

const items = [
	{
		label: 'Item 1',
		value: 'item1',
	}, {
		label: 'Item 2',
		value: 'item2',
	}, {
		label: 'Item 3',
		value: 'item3',
		disabled: true,
	},
];

describe('CheckboxGroup component', () => {
	test('Matches the snapshot', () => {
		const wrapper = shallow(<CheckboxGroup header="Header" items={items} />);
		expect(wrapper).toMatchSnapshot();
	});

	test('OnChange in child triggers parent', () => {
		const onChange = jest.fn();
		const wrapper = mount(<CheckboxGroup header="Header" onChange={onChange} items={items} />);
		wrapper.find(Checkbox).first().find('input').simulate('change', {target: {checked: true}});
		expect(onChange).toBeCalled();
	});
});
