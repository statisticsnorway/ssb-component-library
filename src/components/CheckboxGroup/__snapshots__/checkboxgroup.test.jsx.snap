// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CheckboxGroup component Matches the snapshot 1`] = `
<DocumentFragment>
  <div
    class="ssb-checkbox-group"
  >
    <div
      class="checkbox-group-header"
    >
      Header
    </div>
    <div
      class="boxes flex-column"
    >
      <div
        class="ssb-checkbox"
      >
        <input
          id="item1"
          tabindex="0"
          type="checkbox"
          value="item1"
        />
        <label
          class="checkbox-label"
          for="item1"
        >
          Item 1
        </label>
      </div>
      <div
        class="ssb-checkbox"
      >
        <input
          id="item2"
          tabindex="0"
          type="checkbox"
          value="item2"
        />
        <label
          class="checkbox-label"
          for="item2"
        >
          Item 2
        </label>
      </div>
      <div
        class="ssb-checkbox"
      >
        <input
          disabled=""
          id="item3"
          tabindex="0"
          type="checkbox"
          value="item3"
        />
        <label
          class="checkbox-label"
          for="item3"
        >
          Item 3
        </label>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
